### PR TITLE
Data server

### PR DESCRIPTION
**Pull request by mfleader**
Refactor data capture logic into functions for reuse. Add store function to encapsulate the logic of our three data storage options: local, pbench, and an HTTP data server.

### CODE REVIEW

File: ocp_diagnosis.sh [hunk 1]
```diff
@@ -125,6 +126,23 @@ if [[ "$PROMETHEUS_CAPTURE" == "true" ]]; then
 	fi
 fi
 
-if [[ "$OPENSHIFT_MUST_GATHER" == "true" ]]; then
-	must_gather
+# if [[ "$OPENSHIFT_MUST_GATHER" == "true"]]; then
+# 	must_gather
+# 	curl $DATA_SERVER --form file@"$OUTPUT_DIR/must-gather-$ts.tar.xz"
+# fi
+
```
Comment (mohit-sheth) [inline ocp_diagnosis.sh]: you can remove this

File: ocp_diagnosis.sh [hunk 2]
```diff
@@ -125,6 +126,23 @@ if [[ "$PROMETHEUS_CAPTURE" == "true" ]]; then
...
+# validate data server string is not empty
+if [[ -n "$DATA_SERVER" ]]; then
```
Comment (mohit-sheth) [inline ocp_diagnosis.sh]: we can replace this with
if [[ STORAGE_MODE == data_server]]
  then do stuff
fi
and remove the check inside

File: ocp_diagnosis.sh [hunk 3]
```diff
@@ -125,6 +126,23 @@ if [[ "$PROMETHEUS_CAPTURE" == "true" ]]; then
...
+		curl $DATA_SERVER --form file=@"$OUTPUT_DIR/must-gather-$ts.tar.xz"
+	fi
+
+	if [[ "$PROMETHEUS_CAPTURE" == "true" && "$STORAGE_MODE" == "bottle" ]]; then
+		# tar file has the same name for cases: wal, full
+		curl $DATA_SERVER --form file=@"$OUTPUT_DIR/prometheus-$ts.tar.xz"
+	fi
+
+	else 
+		echo "Data server is undefined. It must be defined to post results to it."
```
Comment (mohit-sheth) [inline ocp_diagnosis.sh]: this can be indented out to match the if statement

**Review by mohit-sheth — CHANGES_REQUESTED** [review]

File: ocp_diagnosis.sh (hunk 2 above)
Comment (mfleader) [reply ocp_diagnosis.sh]: No plans to check `$DATA_SERVER`?

File: ocp_diagnosis.sh [hunk 4]
```diff
@@ -70,51 +70,65 @@ prometheus_pod=$(oc get pods -n $prometheus_namespace | grep -w "Running" | awk
...
 }
 
+
 function must_gather() {
-	echo "================================================================================="
-	echo "                     Running OpenShift must-gather                               "
-	echo "================================================================================="
-	echo "Asssuming the oc client on the host supports collecting must-gather"
-	oc adm must-gather --dest-dir=$OUTPUT_DIR/must-gather-$ts
+	if [[ "$OPENSHIFT_MUST_GATHER" == "true" ]]; then
```
Comment (mohit-sheth) [inline ocp_diagnosis.sh]: this is not necessary

**Comment by mfleader** [comment]
@chaitanyaenr PTAL

**Comment by mohit-sheth** [comment]
@mfleader did an e2e test and it was able to send prom-dump to the server and to pbench as well
lgtm

File: ocp_diagnosis.sh [hunk 5]
```diff
@@ -70,51 +66,66 @@ prometheus_pod=$(oc get pods -n $prometheus_namespace | grep -w "Running" | awk
...
 }
-	
-if [[ "$PROMETHEUS_CAPTURE" == "true" ]]; then
+
+
+function validate_server_is_up() {
+	# return
+	# 	http request code
+
+	curl $STORAGE_MODE 1>/dev/null
```
Comment (chaitanyaenr) [inline ocp_diagnosis.sh]: Maybe set the storage_mode to something like data_server and allow user to pass in the address?

File: ocp_diagnosis.sh [hunk 6]
```diff
@@ -123,8 +134,47 @@ if [[ "$PROMETHEUS_CAPTURE" == "true" ]]; then
...
+		$1;
+	elif [[ $STORAGE_MODE == "pbench" ]]; then
+		# store on the pbench server (STORAGE_MODE=pbench)
+		set_pbench;
+		$1;
+	elif [[ validate_server_is_up -eq 0 ]]; then
+		# store it on the data server (STORAGE_MODE=http://ec2.0.0.0.0:7070)
+		$1 && validate_server_is_up && post_to_server "$OUTPUT_DIR/$2"
+	else
+		echo "Invalid storage mode chosen. STORAGE_MODE is $STORAGE_MODE"
```
Comment (chaitanyaenr) [inline ocp_diagnosis.sh:165]: Maybe exit 1 if the storage mode set is not valid?

Comment (mfleader) [reply ocp_diagnosis.sh:165]: added

File: ocp_diagnosis.sh (hunk 5 above)
Comment (mfleader) [reply ocp_diagnosis.sh]: Changed purpose of STORAGE_MODE added another environment variable to take the data server url, DATA_SERVER_URL

**Review by chaitanyaenr — APPROVED** [review]
LGTM

**Pull request merged**